### ISSUE TITLE
prefetch_data_* empty in arcstats, git repository handling really slow (many small files), zfs_master

### ISSUE THREAD

**Issue by kernelOfTruth** [zfsonlinux/zfs#7477]
Distribution Name       |  Gentoo Linux ~amd64
Distribution Version    |  

emerge --info
Portage 2.3.27 (python 3.5.5-final-0, default/linux/amd64/17.0/desktop, gcc-7.3.0, glibc-2.26-r6, 4.16.3-PDS-0.98n x86_64)

System uname: Linux-4.16.3-PDS-0.98n-x86_64-Intel-R-_Xeon-R-_CPU_E3-1245_v3_@_3.40GHz-with-gentoo-2.3
KiB Mem:    32938404 total,  14013076 free
KiB Swap:   15357948 total,  15357948 free

ZFS Version                  | 
SPL Version                  | 

https://github.com/kernelOfTruth/zfs/commits/zfs_master_19.04.2018
https://github.com/kernelOfTruth/spl/commits/spl_master_17.04.2018

modinfo zfs | grep -iw version **0.7.0-1**
modinfo spl | grep -iw version  **0.7.0-1**

```cat /proc/spl/kstat/zfs/arcstats 
13 1 0x01 98 26656 149074500448 18032414131754
name                            type data
hits                            4    88669478
misses                          4    4340261
demand_data_hits                4    1771721
demand_data_misses              4    240151
demand_metadata_hits            4    82339857
demand_metadata_misses          4    3552945
prefetch_data_hits              4    0
prefetch_data_misses            4    0
prefetch_metadata_hits          4    4557900
prefetch_metadata_misses        4    547165
mru_hits                        4    25733601
mru_ghost_hits                  4    477632
mfu_hits                        4    60522450
mfu_ghost_hits                  4    45945
deleted                         4    527627
mutex_miss                      4    16
access_skip                     4    8
evict_skip                      4    597998
evict_not_enough                4    732
evict_l2_cached                 4    0
evict_l2_eligible               4    24176778752
evict_l2_ineligible             4    1694306304
evict_l2_skip                   4    0
hash_elements                   4    635879
hash_elements_max               4    1068773
hash_collisions                 4    203401
hash_chains                     4    42900
hash_chain_max                  4    6
p                               4    7727199744
c                               4    12884901888
c_min                           4    10737418240
c_max                           4    12884901888
size                            4    10020155944
compressed_size                 4    8741887488
uncompressed_size               4    12245943296
overhead_size                   4    667845120
hdr_size                        4    211232920
data_size                       4    8507392000
metadata_size                   4    902340608
dbuf_size                       4    87314576
dnode_size                      4    230531840
bonus_size                      4    81344000
anon_size                       4    2727936
anon_evictable_data             4    0
anon_evictable_metadata         4    0
mru_size                        4    5443365376
mru_evictable_data              4    4943197696
mru_evictable_metadata          4    41919488
mru_ghost_size                  4    1245714944
mru_ghost_evictable_data        4    268203008
mru_ghost_evictable_metadata    4    977511936
mfu_size                        4    3963639296
mfu_evictable_data              4    3079335936
mfu_evictable_metadata          4    449870336
mfu_ghost_size                  4    3435875840
mfu_ghost_evictable_data        4    131072
mfu_ghost_evictable_metadata    4    3435744768
l2_hits                         4    0
l2_misses                       4    0
l2_feeds                        4    0
l2_rw_clash                     4    0
l2_read_bytes                   4    0
l2_write_bytes                  4    0
l2_writes_sent                  4    0
l2_writes_done                  4    0
l2_writes_error                 4    0
l2_writes_lock_retry            4    0
l2_evict_lock_retry             4    0
l2_evict_reading                4    0
l2_evict_l1cached               4    0
l2_free_on_write                4    0
l2_abort_lowmem                 4    0
l2_cksum_bad                    4    0
l2_io_error                     4    0
l2_size                         4    0
l2_asize                        4    0
l2_hdr_size                     4    0
memory_throttle_count           4    0
memory_direct_count             4    0
memory_indirect_count           4    0
memory_all_bytes                4    33728925696
memory_free_bytes               4    15754919936
memory_available_bytes          3    15227908096
arc_no_grow                     4    0
arc_tempreserve                 4    0
arc_loaned_bytes                4    0
arc_prune                       4    263227
arc_meta_used                   4    1512763944
arc_meta_limit                  4    3221225472
arc_dnode_limit                 4    322122547
arc_meta_max                    4    3299622344
arc_meta_min                    4    16777216
async_upgrade_sync              4    0
demand_hit_predictive_prefetch  4    0
demand_hit_prescient_prefetch   4    0
arc_need_free                   4    0
arc_sys_free                    4    527014464
arc_raw_size                    4    0
```

### Describe the problem you're observing
prefetch_data* is empty in arcstats:

prefetch_data_hits              4    0
prefetch_data_misses            4    0

and disk access also appears unnecessarily slow (e.g. when access and working on git kernel repositories)

### Describe how to reproduce the problem
use a recent zfs_master & spl_master build

In case any of the module options can cause this behavior (potential misconfiguration ?):

**/etc/modprobe.d/zfs.conf**
options zfs zio_dva_throttle_enabled=0
options zfs zfs_arc_max=0x300000000
options zfs zfs_arc_min=0x280000000
options zfs zfs_arc_meta_limit=3221225472
options zfs zfs_dirty_data_max_max_percent=15
options zfs zfs_arc_shrink_shift=9
options zfs spa_slop_shift=6
options zfs zfs_arc_meta_prune=1000000
options zfs ignore_hole_birth=1
options zfs zfs_vdev_async_write_min_active=2
options zfs zfs_delay_scale=8000000
options zfs zfs_vdev_queue_depth_pct=50
options zfs zfs_delay_min_dirty_percent=40
options zfs zfs_metaslab_fragmentation_threshold=80
options zfs zfs_mg_fragmentation_threshold=90

**/etc/modprobe.d/spl.conf**
options spl spl_kmem_cache_kmem_limit=2048
options spl spl_kmem_cache_slab_limit=16384
options spl spl_kmem_cache_expire=2
options spl spl_kmem_cache_reclaim=1
options spl spl_kmem_cache_max_size=64
options spl spl_kmem_cache_obj_per_slab=16

**Comment by dweeezil** [comment]
@kernelOfTruth Thanks for the report.  I have a hunch this was broken by the device evacuation work and will take a look at it today.

**Comment by dweeezil** [comment]
Yes, indeed, it would certainly appear that the device evacuation work erroneously will disable prefetch for pools for which no device evacuation has actually occurred.  This is most likely a problem upstream, too.  I'll work up a fix for it.

**Comment by dweeezil** [comment]
I've worked up a3f7e7d9 as a possible fix.  @ahrens, what are your thoughts on this?

**Comment by ahrens** [comment]
Please see https://github.com/openzfs/openzfs/pull/607/files.  That should be upstreamed very soon.

**Comment by dweeezil** [comment]
Thanks.  I did check upstream commits but didn't check PRs.  Oops.

**Comment by prakashsurya** [comment]
merged: https://github.com/openzfs/openzfs/commit/16127b627bbb36d736aa7de17859fe8444fc0cce

**Comment by behlendorf** [comment]
Closing.  This fix was applied, 20507534d4ede14d4dd82c99fc8d461704ce7419.
